File: Chapter16 - Moderate/Q7-NumberMax.cpp
#include <iostream>
#include <climits>

using namespace std;

int getSignOf(int n){
    return ((n >> 31) & 1) ^ 1;
}

int numberMax(int a, int b){
    int sa = getSignOf(a);
    int sb = getSignOf(b);
    int sc = getSignOf(a-b);

    int differentSigns = sa ^ sb;

    int useAsign = sa ^ sb;
    int useCsign = useAsign ^ 1;

    int k = useAsign * sa + useCsign * sc;

    int q = k ^ 1;
    return a*k + b*q;
}

int main(){
    cout << numberMax(1,2)<< endl;
    cout << numberMax(-1,5) << endl;
    cout << numberMax(3,3) << endl;
    cout << numberMax(0,0)<< endl;
    cout << numberMax(13,0)<< endl;
    cout << numberMax(-5,-2) << endl;
    cout << numberMax(INT_MAX -2 , -15) << endl;
    return 0;
}

File: Chapter16 - Moderate/Q7-NumberMax.py
def getSignOf(n):
    return (n >> 31) & 1

def numberMax(n1, n2):
    s1 = getSignOf(n1)
    s2 = getSignOf(n2)
    s3 = getSignOf(n1 - n2)
    k = getSignOf(n1-n2)
    q = k ^ 1
    return n1*q + n2*(1-q)

def main():
    print(numberMax(1,2))
    print(numberMax(-1,5))
    print(numberMax(3,3))
    print(numberMax(0,0))
    print(numberMax(13,0))
    print(numberMax(-5,-2))

main()